File: Archive-2015/10. GesturesLiveDemo/GesturesLiveDemo/ViewController.m
//
//  ViewController.m
//  GesturesLiveDemo
//
//  Created by Doncho Minkov on 11/3/14.
//  Copyright (c) 2014 Telerik Academy. All rights reserved.
//

#import "ViewController.h"
#import <CoreGraphics/CoreGraphics.h>
#import "RectView.h"

@interface ViewController ()

@property (nonatomic, strong) RectView* rectView;

@property CGFloat rotationAngle;
@property CGFloat scaleFactor;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.scaleFactor = 1;
    self.rotationAngle = 0;
    
    NSInteger side = 60;
    
    NSInteger x = self.view.bounds.origin.x + (self.view.bounds.size.width - side)/2;
    NSInteger y = self.view.bounds.origin.y + (self.view.bounds.size.height - side)/2;
    
    CGRect rect = CGRectMake(x, y, side, side);
    
    RectView* rectView = [[RectView alloc] initWithFrame:rect];
    
    [self.view addSubview:rectView];
    
    rectView.transform = CGAffineTransformMakeScale(1, 1);
    
    self.rectView = rectView;
    
    //    UISwipeGestureRecognizer *rightSwipeRecognizer = [[UISwipeGestureRecognizer alloc] initWithTarget:self
    //                                                                                               action:@selector(swipe:)];
    //
    //    UISwipeGestureRecognizer *leftSwipeRecognizer = [[UISwipeGestureRecognizer alloc] initWithTarget:self
    //                                                                                              action:@selector(swipe:)];
    //    leftSwipeRecognizer.direction = UISwipeGestureRecognizerDirectionLeft;
    //
    //    [self.view addGestureRecognizer:rightSwipeRecognizer];
    //    [self.view addGestureRecognizer:leftSwipeRecognizer];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (IBAction)swipe:(UISwipeGestureRecognizer *)recognizer {
    NSString *message;
    switch (recognizer.direction) {
        case UISwipeGestureRecognizerDirectionRight:
            message = @"Swiped right";
            break;
        case UISwipeGestureRecognizerDirectionLeft:
            message = @"Swiped left";
            break;
        default:
            message =@"Not swiped";
            break;
    }
    NSLog(@"%@", message);
}

- (IBAction)pinch:(UIPinchGestureRecognizer *)sender {
    //initial scale 1
    //    CGFloat scaleFactor = self.scaleFactor * sender.scale;
    CGFloat scaleFactor = sender.scale;
    
    //    CGAffineTransform currentTrans = self.rectView.transform;
    //
    CGAffineTransform transform = CGAffineTransformMakeScale(scaleFactor, scaleFactor);
    
    self.scaleFactor = scaleFactor;
    
    self.rectView.transform = CGAffineTransformRotate(transform, self.rotationAngle);
}

- (IBAction)rotate:(UIRotationGestureRecognizer *)sender {
    //initial 0
    
    self.rotationAngle = sender.rotation;
    
    CGAffineTransform transform = CGAffineTransformMakeScale( self.scaleFactor, self.scaleFactor);
    
    self.rectView.transform = CGAffineTransformRotate(transform, self.rotationAngle);
}

- (IBAction)pan:(UIPanGestureRecognizer *)sender {
    CGPoint panLocation = [sender locationInView:self.view];
    
    CGRect rectViewBounds = self.rectView.frame;
    
    if(![self panAt:panLocation isInRect: rectViewBounds]){
        return;
    }
    
    CGFloat offsetX = -self.view.bounds.size.width/2;
    CGFloat offsetY = -self.view.bounds.size.height/2;
    
    CGFloat tx = self.scaleFactor * (panLocation.x + offsetX);
    CGFloat ty = self.scaleFactor * (panLocation.y + offsetY);
    
    self.rectView.transform = CGAffineTransformIdentity;
    
    CGAffineTransform transform = CGAffineTransformMakeScale(self.scaleFactor, self.scaleFactor);
    transform = CGAffineTransformRotate(transform, self.rotationAngle);
    transform = CGAffineTransformTranslate(transform, tx, ty);
    
    self.rectView.transform = transform;
}

-(BOOL) panAt: (CGPoint) point isInRect: (CGRect) rect{
    CGFloat margin = 15;
    CGFloat top = rect.origin.y - margin;
    CGFloat bottom = rect.origin.y + rect.size.height + margin;
    CGFloat left = rect.origin.x - margin;
    CGFloat right = rect.origin.x + rect.size.width + margin;
    return (left <= point.x && point.x <= right) &&
    (top <= point.y && point.y <= bottom);
}
@end

File: demos/GesturesDemos/GesturesDemos/ViewController.swift
//
//  ViewController.swift
//  GesturesDemos
//
//  Created by Doncho Minkov on 3/29/17.
//  Copyright © 2017 Doncho Minkov. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    var rect: UIView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.rect = UIView(frame: CGRect(x: self.view.frame.width / 2,
                                         y: self.view.frame.height / 2,
                                         width: 50, height: 50))
        self.rect?.backgroundColor = .blue
        self.view.addSubview(self.rect!)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    @IBAction func move(_ panGesture: UIPanGestureRecognizer) {
        let location = panGesture.location(in: self.view)
        if !self.isInRect(location) {
            return
        }
        
        self.rect?.frame.origin = CGPoint(x: location.x - 25,
                                          y: location.y - 25)
    }
    
    var scale: CGFloat = 1.0
    
    var isTouch = false
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        isTouch = true
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        print(touches)
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        isTouch = false
    }
    
    @IBAction func scale(_ pinchGesture: UIPinchGestureRecognizer) {
        var newScale = pinchGesture.scale
            
        if !self.isTouch {
            newScale = newScale * self.scale
            self.scale = newScale
            self.isTouch = true
        }
        
        self.rect?.transform =  CGAffineTransform(scaleX: newScale, y: newScale)
    }
    
    
    @IBAction func rotate(_ rotateGesture: UIRotationGestureRecognizer) {
        let angle = rotateGesture.rotation
        
        self.rect?.transform = CGAffineTransform(rotationAngle: angle)
    }
    
    func isInRect(_ position: CGPoint) -> Bool {
        return (self.rect?.frame.minX)! <= position.x &&
            position.x <= (self.rect?.frame.maxX)! &&
            (self.rect?.frame.minY)! <= position.y &&
            position.y <= (self.rect?.frame.maxY)!
    }
    
}

